### ISSUE TITLE
Window resize is slightly broken

### ISSUE THREAD

**Issue by equalsraf** [equalsraf/neovim-qt#4]
(in the tb-remote-ui branch) resizing the GUI may have some unexpected effects e.g.

1. We drag the window to resize and afterwards the window is resized back
1. We resize the window, but the actual Neovim work area remains the same size, with a wide margin around it

There are multiple reasons for this issue

1. try_resize_ui does not guarantee a resize and SHOULD NOT be called simultaneously - Neovim will ignore paralell calls
1. When the user is dragging the window it will cause multiple calls to try_resize_ui, emphasizing the previous point
1. resizing can be expensive in drawing operations (i.e. takes time)

There are ways around these issues, but I would prefer to wait until the Neovim API becomes a bit more stable and the UI functions are place in the API metadata object (so the bindings generate signals for those calls as well). Preliminary suggestions

1. Collapse multiple calls to try_resize_ui and only make a single call at a time - this is similar to what was done for vim-qt
1. Delay repainting due to window resizing.

**Comment by equalsraf** [comment]
Have an ongoing fix for this, but I'm having some issues with this https://github.com/equalsraf/neovim-qt/issues/4

**Comment by equalsraf** [comment]
[QTBUG-45086](https://bugreports.qt.io/browse/QTBUG-45806)

**Comment by equalsraf** [comment]
#28 was a big help but caused an issue when using i3 as window manager. Because we try to attach the Neovim window as 2/3 of the available screen space - but the tilled window manager does not allow the window to be resized - this causes a margin to appear around the Neovim widget.

![bug1](https://cloud.githubusercontent.com/assets/715238/8633120/cca6b67a-27b1-11e5-9ebb-e8592e336923.png)

The margin background is QTBUG-45086. But the actual size issue can be fixed. We need to treat that default size a bit more carefully, e.g.

```diff

--- a/src/gui/shell.cpp
+++ b/src/gui/shell.cpp
@@ -147,8 +147,7 @@ void Shell::neovimIsReady()
        connect(m_nvim->neovimObject(), &Neovim::neovimNotification,
                        this, &Shell::handleNeovimNotification);
        // FIXME: this API will change
-       QRect screenRect = QApplication::desktop()->availableGeometry(this);
-       m_nvim->attachUi(screenRect.width()*0.66/neovimCellWidth(), screenRect.height()*0.66/neovimRowHeight());
+       m_nvim->attachUi(width()/neovimCellWidth(), height()/neovimRowHeight());
 
        connect(m_nvim->neovimObject(), &Neovim::on_ui_try_resize,
                        this, &Shell::neovimResizeFinished);
```

defaulting to the current widget size seems to work well for this case, but I can't test on other WM (or Windows) right now.

**Comment by timshannon** [comment]
Other than this issue this seems to be the best GUI client for neovim I've found on linux.

Is there any sort of workaround for this?

**Comment by equalsraf** [comment]
@timshannon, sadly no workaround, the only way to fix it is probably to rewrite the shell widget to avoid resizes, or fix the issues upstream.

**Comment by timshannon** [comment]
@equalsraf upstream QT or upstream neovim?  If it's the latter, do you have an issue # I can follow?

**Comment by equalsraf** [comment]
Upstream Qt
- https://bugreports.qt.io/browse/QTBUG-45806
- https://bugreports.qt.io/browse/QTBUG-39686 (we don't need this one for resize, but would be nice for optimisations)

**Comment by equalsraf** [comment]
#109 has been merged, we can move this discussion into #108 TLDR is neovim-qt resizes the window, ignoring instructions from Neovim. The margin issues discussed above should be fixed.

**Comment by timshannon** [comment]
Works great :+1:
